net-utils/src/token_bucket.rs

```diff
@@ -0,0 +1,360 @@
+    #[allow(clippy::arithmetic_side_effects)]
+    pub fn new(target_capacity: usize, prototype_bucket: TokenBucket, shard_amount: usize) -> Self {
+        Self {
+            data: DashMap::with_capacity_and_shard_amount(target_capacity * 2, shard_amount),
+            target_capacity,
+            prototype_bucket,
+            countdown_to_shrink: AtomicUsize::new(0),
+            approx_len: AtomicUsize::new(0),
+        }
+    }
+
+    /// Fetches amount of tokens available for key.
+    ///
+    /// Returns None if no bucket exists for the key provided
+    #[inline]
+    pub fn current_tokens(&self, key: impl Borrow<K>) -> Option<u64> {
+        let bucket = self.data.get(key.borrow())?;
+        Some(bucket.current_tokens())
+    }
+
+    /// Consumes request_size tokens from a bucket at given key.
+    ///
+    /// On success, returns Ok(amount of tokens left in the bucket)
+    /// On failure, returns Err(amount of tokens missing to fill request)
+    /// If no bucket exists at key, a new bucket will be allocated, and normal policy will be applied to it
+    /// Outdated buckets may be evicted on an LRU basis.
+    pub fn consume_tokens(&self, key: K, request_size: u64) -> Result<u64, u64> {
+        let (entry_added, res) = {
+            let bucket = self.data.entry(key);
+            match bucket {
+                Entry::Occupied(entry) => (false, entry.get().consume_tokens(request_size)),
+                Entry::Vacant(entry) => {
+                    // if the key is not in the LRU, we need to allocate a new bucket
+                    let bucket = self.prototype_bucket.clone();
+                    let res = bucket.consume_tokens(request_size);
+                    entry.insert(bucket);
+                    (true, res)
+                }
+            }
+        };
+
+        if entry_added {
+            // we want to check for length, but not too often
+            // this should reduce probability of lock contention
+            let step = self.target_capacity / 4;
+            if self
+                .countdown_to_shrink
+                .fetch_update(Ordering::Relaxed, Ordering::Relaxed, |v| {
+                    if v >= step {
+                        Some(0)
+                    } else {
+                        Some(v.saturating_add(1))
+                    }
+                })
+                .unwrap_or_default()
+                >= step
+            {
+                self.maybe_shrink();
+            }
+        }
+        res
+    }
+
+    /// Returns approximate amount of entries in the datastructure.
+    /// Should be within ~10% of the true amount.
+    #[inline]
+    pub fn len_approx(&self) -> usize {
+        self.approx_len.load(Ordering::Relaxed)
+    }
+
+    // apply lazy-LRU eviction policy to each DashMap shard.
+    // Allowing side-effects here since overflows here are not
+    // actually possible
+    #[allow(clippy::arithmetic_side_effects)]
+    fn maybe_shrink(&self) {
```

Comment:
I am concerned many threads doing the same thing when it is near the condition, contending unnecessarily, maybe we should have a pub interface and let a single maintainer thread to do the period cleaning.

Reply:
This is why there is logic to only call it when countdown_to_shrink is 0.